kernel/of/fdt.c

```diff
@@ -359,8 +494,240 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
         if (fdt_node_is_compatible(fdt, node, compat)) {
             return node;
         }
-        node = fdt_next_node(fdt, node);
+        node = fdt_next_node(fdt, node, NULL);
+    }
+
+    return NULL;
+}
+
+struct fdt_node*
+fdt_find_node_by_device_type(struct fdt* fdt, struct fdt_node* start, const char* type)
+{
+    struct fdt_node* node;
+    if (start != NULL) {
+        node = fdt_next_node(fdt, start, NULL);
+    }
+    else {
+        node = fdt_node_begin(fdt);
+    }
+
+    size_t len = (size_t)strlen(type);
+    while (node != NULL) {
+        struct fdt_prop* prop = fdt_get_prop_by_name(fdt, node, NULL, "device_type");
+        if (prop != NULL) {
+            size_t prop_len = fdt_prop_val_len(prop);
+            void* prop_val = fdt_prop_val(prop);
+            // Consider equal even if '\0' is omitted in the property
+            if (prop_len == len + 1 || prop_len == len) {
```

Comment:
```suggestion
            if (prop_len == len - 1 || prop_len == len) {
```

Having `\0` omitted would imply `prop_len` is 1 char shorter

Reply:
`len` is from `strlen` which already omits the `\0`, so the first case is for it the prop doesn't omit the `\0`

Reply:
icic ok